contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,325 @@
+//SPDX-License-Identifier: MIT
+pragma solidity >= 0.6.0 < 0.7.0;
+
+import "./access/Ownable.sol";
+import "./utils/Pausable.sol";
+
+import {SafeMath} from "./math/SafeMath.sol";
+
+contract RockPaperScissors is Ownable, Pausable {
+
+    using SafeMath for uint256;
+
+    mapping(address => uint256) public balances;
+    mapping(bytes32 => Game) public games;
+
+    address private constant NULL_ADDRESS = address(0);
+    string private constant INVALID_PLAYER_MSG = "Invalid player";
+    string private constant INVALID_MOVE_MSG = "Invalid move";
+    string private constant INVALID_STEP_MSG = "Invalid step";
+    string private constant HASH_MISMATCH_MSG = "Move and secret do not match";
+    string private constant GAME_NOT_EXPIRED_MSG = "Game has not expired";
+    string private constant SECRET_EMPTY_MSG = "Secret is empty";
+
+    uint256 constant FORFEIT_WINDOW = 1 days;
+    uint256 constant FEE_PERCENTAGE = 10;
+
+    enum Steps {
+        INIT, // 0
+        PLAYER_ONE_MOVE, // 1
+        PLAYER_TWO_MOVE, // 2
+        PLAYER_ONE_REVEAL // 3
+    }
+
+    Steps public steps;
+
+    struct Game {
+        uint256 stake;
+        uint256 expiryDate;
+        uint256 playerOneMove;
+        uint256 step;
```

Comment:
I found out that uint8 uses more gas than uint256.

https://ethereum.stackexchange.com/questions/3067/why-does-uint8-cost-more-gas-than-uint256

Reply:
Overall for a given game, or for a specific tx?
Because those 2 `uint8` are next to each other so they would use a single storage slot. Perhaps that will provide saving when you go from `0` -> non-`0` only once for this slot, instead of twice.
Plus, at erase time, you can use the assembly optimisation to clear the slot (the 2 variables) in 1 `SSTORE`.
That's why I ask if it is the whole game that is cheaper or a single tx.

Reply:
OK. I see. It relates to our conversation on Monday. I have reverted and placed next to each other. I haven't done a gas comparison by function or whole game but something I can consider. Perhaps a micro-optimisation given such a small difference in gas.